crates/accelerate/src/split_2q_unitaries.rs

```diff
@@ -76,12 +129,78 @@ pub fn split_2q_unitaries(
                 dag.replace_node_with_1q_ops(py, node, insert_fn)?;
                 dag.add_global_phase(py, &Param::Float(decomp.global_phase))?;
             }
-            // TODO: also look into splitting on Specialization::Swap and just
-            // swap the virtual qubits. Doing this we will need to update the
-            // permutation like in ElidePermutations
         }
     }
-    Ok(())
+    if !split_swaps || !has_swaps {
+        return Ok(None);
+    }
+    // We have swap-like unitaries, so we create a new DAG in a manner similar to
+    // The Elide Permutations pass, while also splitting the unitaries to 1-qubit gates
+    let mut mapping: Vec<usize> = (0..dag.num_qubits()).collect();
+    let mut new_dag = dag.copy_empty_like(py, "alike")?;
+    for node in dag.topological_op_nodes()? {
+        if let NodeType::Operation(inst) = &dag.dag()[node] {
+            let qubits = dag.get_qargs(inst.qubits).to_vec();
+            let mut is_swap = false;
+            if qubits.len() == 2 && inst.op.name() == "unitary" {
+                let matrix = inst
+                    .op
+                    .matrix(inst.params_view())
+                    .expect("'unitary' gates should always have a matrix form");
+                let decomp = TwoQubitWeylDecomposition::new_inner(
+                    matrix.view(),
+                    Some(requested_fidelity),
+                    None,
+                )?;
+                if matches!(decomp.specialization, Specialization::SWAPEquiv) {
+                    // perform the virtual swap
+                    is_swap = true;
+                    let qargs = dag.get_qargs(inst.qubits);
+                    let index0 = qargs[0].index();
+                    let index1 = qargs[1].index();
+                    mapping.swap(index0, index1);
+                    // now add the two 1-qubit gates
+                    let (k1r_gate, k1l_gate) = create_k1_gates(&decomp, py)?;
+                    add_new_op(
+                        &mut new_dag,
+                        k1r_gate.extract()?,
+                        vec![qubits[0]],
+                        &mapping,
+                        py,
+                    )?;
+                    add_new_op(
+                        &mut new_dag,
+                        k1l_gate.extract()?,
+                        vec![qubits[1]],
+                        &mapping,
+                        py,
+                    )?;
+                    new_dag.add_global_phase(py, &Param::Float(decomp.global_phase + PI4))?;
+                }
+            }
+            if !is_swap {
```

Comment:
Instead of doing this with a mutable `is_swap` boolean variable, couldn't you just add a `continue` inside the if block for a swap equivalent unitary? and remove this `!is_swap` if statement. I feel like that might be a bit preferable because it reduce a layer of indentation here if nothing else. Although I suspect the compiler might be able to reason about the flow a bit better with a continue, but that would need to be validated and if true is of very minimal relevance to not be worth validating or benchmarking.

Reply:
Done in a9bc3e1ea